### PR TITLE
Add unstable cfg_if! macro to libcore

### PR DESCRIPTION
**Pull request by gnzlbg**
This PR adds the `cfg_if!` macro to libcore and exposes it behind the `cfg_if` feature-gate. 

Is this macro useful ? It has [>140 reverse dependencies on crates.io](https://crates.io/crates/cfg-if/reverse_dependencies). 

Pros of exposing it from `libcore`: in `libcore` we can't use the crates.io version unless we stabilize the `no_core` feature, which is something we might not want to do. The macro is already there - in fact, it's there twice (once in internal_macros, once as part of core::arch) - so we might as well expose it once, so that everybody can just easily use it. Note that even though libstd can technically use `cfg_if` from crates.io, to minimize external dependencies, the macro is copy-pasted another 3-4 times (std_arch, std_detect, hashbrown, libstd-itself, ...). 

Cons of exposing it from `libcore`: extra maintenance burden, if we were to stabilize `no_core`, `libcore` could just use a dependency from crates.io, the cost of copy-pasting it N times and maintaining the copies is less work than exposing it from libcore. Also, it might well be that some day we might want to solve this problem using language features (e.g. allowing people to programatically create their own `cfg`s, etc.). If that ever happens, we can just deprecate it.

r? @SimonSapin cc @rust-lang/libs

### CODE REVIEW

File: src/libcore/macros.rs [hunk 1]
```diff
@@ -635,6 +635,89 @@ macro_rules! uninitialized_array {
...
+///     } else if #[cfg(target_pointer_width = "32")] {
+///         fn foo() { /* non-unix, 32-bit functionality */ }
+///     } else {
+///         fn foo() { /* fallback implementation */ }
+///     }
+/// }
+///
+/// # fn main() {}
+/// ```
+#[macro_export(local_inner_macros)]
```
Comment (Centril) [inline src/libcore/macros.rs]: Use `$crate::` instead?

**Comment by SimonSapin** [comment]
Thanks @gnzlbg! I’m in favor of this PR, and there’s precedent for accepting unstable APIs without an RFC. However this adds a new macro **to the libcore prelude**, so it deserves special attention from @rust-lang/libs. In any case, we’ll at least do a FCP before stabilizing.

Given there there is no concrete proposal so far for stabilizing `no_core` and that IMO it would only have very limited usefulness (`cfg-if` is the only crate I can think of), I don’t think it’s worth making hypothetical scenarios based on this.

> Cons of exposing it from `libcore`: extra maintenance burden

On the contrary I think we can reduce maintenance burden by having fewer copies of the macro. Could you add commits to this PR to remove the ones in `src/libstd/macros.rs`, `src/libpanic_unwind/macros.rs`, and `src/libunwind/macros.rs`? Maybe also changing the `librustc_data_structures` and `libsyntax_pos` crates to use it from libcore rather than crates.io (on `cfg(not(stage0))`).

It’d also be nice to either add a new test or point to an existing test to show that when an unstable macro is in the prelude, it can be shadowed by a stable macro with the same name from another crate without the crate using the macro opting into any feature gate. (The case relevant here is using crates.io’s `cfg-if` in Rust 1.35 on the Stable channel.)

Finally, it looks like the multiple existing implementations of this macro are not identical. Could you say if the differences are significant, and why pick this one?

**Comment by gnzlbg** [comment]
> Could you add commits to this PR to remove the ones in src/libstd/macros.rs, src/libpanic_unwind/macros.rs, and src/libunwind/macros.rs?

Is that also ok if I do this once this lands in the next bootstrap compiler? Otherwise one probably still needs to keep the older macros around, at least for bootstraping right ?

**Comment by gnzlbg** [comment]
> Finally, it looks like the multiple existing implementations of this macro are not identical. Could you say if the differences are significant, and why pick this one?

Older ones are probably copy-pastes from older versions of the cfg-if crate (at least that's the case for the libstd and stdsimd ones). I am not sure what bugs the different releases of cfg-if have fixed, but this implementation is the one from its latest release.

**Comment by gnzlbg** [comment]
> It’d also be nice to either add a new test or point to an existing test to show that when an unstable macro is in the prelude, it can be shadowed by a stable macro with the same name from another crate without the crate using the macro opting into any feature gate. (The case relevant here is using crates.io’s cfg-if in Rust 1.35 on the Stable channel.)

I'll add such a test to this PR :)

**Comment by SimonSapin** [comment]
> Otherwise one probably still needs to keep the older macros around, at least for bootstraping right ?

I *think* not: all crates in the repo are compiled against libcore in the repo.

It’s only standard library crates, in stage 0, that are compiled against the bootstrap *compiler*. So they may need `#[cfg]`s for lang item changes. I believe this is the current status, until something like https://github.com/rust-lang/rust/pull/49119 lands. Try it?

**Comment by alexcrichton** [comment]
@SimonSapin should be correct in that all the existing `cfg_if` definitions throughout the codebase should be safely deletable, no bootstrap concerns necessary. 

@SimonSapin as to the litany of definitions most of them probably stem from me as I've copied it to various places over time as I got fed up with the jungle of `#[cfg]` mainenance. Changes to the macro itself over time have been, I think in order:

1. Macro was incepted
2. Final `else { ... }` block was made optional
3. Compatibility importing the macro through the module system and the 2018 edition
4. Adding coments explaining what's going on

Each of the macros may represent some snapshot in time amongst all those, but the general gist is that it's basically been the same over time, the surface syntax never changing except for a final `else` becoming optional.

File: src/libcore/tests/cfg_if_override.rs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+//! Test that overriding cfg_if with our own cfg_if macro does not break
+//! anything.
+#![feature(cfg_if)]
```
Comment (alexcrichton) [inline src/libcore/tests/cfg_if_override.rs]: This directive shouldn't be required, right?

File: src/libstd/lib.rs [hunk 3]
```diff
@@ -241,6 +241,7 @@
 #![feature(bind_by_move_pattern_guards)]
 #![feature(box_syntax)]
 #![feature(c_variadic)]
+#![feature(cfg_if)]
```
Comment (alexcrichton) [inline src/libstd/lib.rs]: I think that the `pub use` from core will have to be updated below to include `cfg_if` as well

File: src/libcore/macros.rs [hunk 4]
```diff
@@ -635,6 +635,89 @@ macro_rules! uninitialized_array {
...
+/// emitting the implementation which matches first.
+///
+/// This allows you to conveniently provide a long list `#[cfg]`'d blocks of code
+/// without having to rewrite each clause multiple times.
+///
+/// # Example
+///
+/// ```
+/// #[macro_use]
+/// extern crate cfg_if;
```
Comment (alexcrichton) [inline src/libcore/macros.rs]: I think you'll want to delete this

File: src/libcore/Cargo.toml [hunk 5]
```diff
@@ -21,6 +21,7 @@ path = "../libcore/benches/lib.rs"
 
 [dev-dependencies]
 rand = "0.6"
+cfg_if = "1.0"
```
Comment (alexcrichton) [inline src/libcore/Cargo.toml]: s/_/-/

File: src/libcore/tests/cfg_if_override.rs (hunk 2 above)
Comment (gnzlbg) [reply src/libcore/tests/cfg_if_override.rs]: I think so. It should only be required if the `cfg_if!` from `core` is used, but that shouldn't happen here.

**Comment by eaglgenes101** [comment]
I would argue that a `cfg_match!` macro makes more sense, as it would parallel its base syntax better than `cfg_if`, which, because macros' inability to look ahead outside of its own body, requires a second layer of expressions for the actual conditional compilation functionality: 
```
cfg_match! {
    #[cfg(unix)] => { fn foo() { /* unix specific functionality */ } },
    #[cfg(target_pointer_width = "32")] => { fn foo() { /* non-unix, 32-bit functionality */ } }, 
    _ => { fn foo() { /* fallback implementation */ } }
}
```

**Comment by Centril** [comment]
@eaglgenes101 Oooh! I had thought of suggesting that also. Cool.

One thing I dislike about `cfg_if! { ... }` is that it looks like `cfg_if! { if `  <-- see the two ifs repeated? That's quite awkward.

**Comment by gnzlbg** [comment]
@eaglgenes101 is there a crate implementing such a macro? How many crates
are using it?

 Note that this PR does not stabilize anything, it just exposes to users
something that a big part of the ecosystem is using. There will be an FCP /
mini-RFC before stabilization, and until that happens there is enough time
for any interested party to come up with a better solution, and convince
the ecosystem to use it. Adding a cfg match later on is not incompatible
with adding cfg if before that happens.

**Comment by faern** [comment]
Regarding the `cfg_match!` syntax mentioned above. It got me thinking. It would be extremely handy if the macro could support single expression branches, like normal matches. So it can be used much more compact with oneliners, like this:
```rust
const PLATFORM_NAME: &str = cfg_match! {
    #[cfg(unix)] => "*nix",
    #[cfg(windows)] => "Win",
    _ => compile_error!("Unsupported platform!"),
};
```
That would reduce sooo much duplication and risk for mistakes by not having to repeat `const PLATFORM_NAME: &str = `

**Comment by Dylan-DPC** [comment]
ping from triage @gnzlbg you have failing tests to resolve. Any updates?

**Comment by jonas-schievink** [comment]
Ping from triage @gnzlbg, this needs merge conflicts to be resolved. I'm not sure what the status of this is – does this still need an RFC/FCP, or is `cfg_match!` clearly a better alternative? cc @rust-lang/libs

**Comment by alexcrichton** [comment]
I've opened https://github.com/rust-lang/rust/pull/61720 to separate out the question of adding an unstable macro from the internal refactorings done here.

**Comment by gnzlbg** [comment]
> does this still need an RFC/FCP, or is cfg_match! clearly a better alternative?

That question is still unresolved.

**Comment by gnzlbg** [comment]
Also, resolving this question would require for somebody to implement a `cfg_match!` macro that works. AFAICT this is not possible in Rust today, because `_` is a pseudo-identifier, and `meta` matches it (see https://github.com/rust-lang/rust/issues/59881). So this is kind of blocked on whether we can improve the macro system to support `_` properly, and then blocked on use such improved macro system to implement a `cfg_match` crate, and then blocked on resolving whether that crate is better in all ways than `cfg_if!`, or on whether we should add both `cfg_if` and `cfg_match`.

**Comment by gnzlbg** [comment]
@rust-lang/libs should decide what to do here. Should this be closed until either someone builds a good `cfg_match!` library or that's proven impossible ? Should we proceed with adding `cfg_if!` to libcore ? Even if its kept perma unstable forever at least it can be reused throughout the internal crates.

**Comment by alexcrichton** [comment]
I would personally prefer that we not take this route. I don't think a libs team decision should be guided by internal implementation details, and for example there are now currently no duplications of the `cfg_if` macro definition in this codebase. While I do think this is a very useful macro it has nowhere near the number of reverse dependencies of a crate like serde, so I don't think it's really that ubiquitous. For that reason I think it's fine to leave to an external crate for now.

**Comment by gnzlbg** [comment]
Does `core::arch` now get `cfg_if!` from `libcore` somehow, or how does that work?

**Comment by alexcrichton** [comment]
It does not get it because libcore doesn't have a definition. It also doesn't currently use it.

**Comment by gnzlbg** [comment]
Indeed, only std_detect needs it but that can be fetched from crates.io.

**Comment by BurntSushi** [comment]
> While I do think this is a very useful macro it has nowhere near the number of reverse dependencies of a crate like serde

FWIW, there have been at least a few occasions where I would have used `cfg_if!` in a library if it were provided by `std`, but have consciously chosen to avoid using it just to avoid bringing in another dep.

(I'm not saying this should necessarily change anything here, but this is one of those things that falls into the category of "really nice if it's available, but maybe not worth the trouble of bringing in another dep in less complex scenarios.)

**Comment by gnzlbg** [comment]
@BurntSushi When I opened this the reason to not go through an RFC was that the design here was small and constrained enough that a tracking issue / FCP might be enough. Some people have raised the issue that a `cfg_match!` would be a better fit, which kind of shows that the design space here is larger than at least I originally thought, and maye that means that adding a macro like this to libcore should go through the RFC process.

**Pull request closed without merging**